### ISSUE TITLE
`files.get` interface differs significantly between Fog::Storage::GoogleJson and Fog::Storage::GoogleXML

### ISSUE THREAD

**Issue by tcdowney** [fog/fog-google#291]
We were trying to consume the new JSON API functionality provided in the latest version (`1.0.1`) of this gem. 

Calls to `files.get` that use the new `GoogleJson` version of `get_object` are not working with our existing code that passes the `files.get` method a block that we use to write the file out to disk. You can see an example of this [here](https://github.com/cloudfoundry/cloud_controller_ng/blob/d9a5c6a2152645749de6996b7fef76dd38b2be90/lib/cloud_controller/blobstore/fog/fog_client.rb#L45-L48).

In the old `GoogleXML` implementation of this, there is apparent [handling of the passed block](https://github.com/fog/fog-google/blob/031a9a941ebde6fd4bf6ba11d3d8a5711a6233c9/lib/fog/storage/google_xml/requests/get_object.rb#L43), whereas in the the new `GoogleJson` implementation [it appears to just return the object directly](https://github.com/fog/fog-google/blob/031a9a941ebde6fd4bf6ba11d3d8a5711a6233c9/lib/fog/storage/google_json/requests/get_object.rb#L65) regardless of whether a block was provided in the call to [files.get](https://github.com/fog/fog-google/blob/031a9a941ebde6fd4bf6ba11d3d8a5711a6233c9/lib/fog/storage/google_json/models/files.rb#L36).

Comments in the new code imply that we need to call this method differently and supply a `download_dest` option.

Unfortunately, this seems like an inconsistency between the two implementations and makes it more difficult to interoperate with `fog-aws` which [functions similarly](https://github.com/fog/fog-aws/blob/d828888b7a4dc3374382090afa16839442aa6d53/lib/fog/aws/requests/storage/get_object.rb#L53) to the `GoogleXML` version. It also makes it very difficult to interoperate with both the `GoogleXML` and `GoogleJson` versions themselves since it's not clear to consumers which codepath the gem will end up using. In our application we allow users to provide arbitrary fog connection config so we want to be able to rely on a standardized interface.

Thanks!
Tim Downey && @ericpromislow

**Comment by emilymye** [comment]
I'll get a fix for handling that block correctly. get_object doesn't actually expect download_dest to be passed in (it's set to a temporary buffer), and the file value is returned as object[:body].

Also, @icco to confirm but I don't think in general the GoogleXML library is being supported anymore. It certainly was not upgraded.

**Comment by icco** [comment]
GoogleXML is still suported, it just uses a completely different backend. GoogleJSON supports https://cloud.google.com/storage/docs/json_api/v1/ and GoogleXML supports https://cloud.google.com/storage/docs/xml-api/overview

**Comment by icco** [comment]
Also @tcdowney could you provide an example of how you're calling the code? We'd love to add it to our tests or examples / use it to debug.

**Comment by emilymye** [comment]
Yeah, it'd be helpful to see the block you're passing in.

@icco it looks like the block was removed [in this pass](https://github.com/fog/fog-google/commit/5b654aad197726f8b0a56e3623b7e49228304904#diff-56862befe007db24133d55e97110864c). I think it was reasonable since the block is excon related. If we want to keep it compatible, we'll probably have to wrap the excon response block in an IO stream to pass to the Google request ([here](https://github.com/google/google-api-ruby-client/blob/master/generated/google/apis/storage_v1/service.rb#L1805)). Does that sound reasonable?

**Comment by icco** [comment]
@emilymye I'm confused, so GoogleXML uses excon, but GoogleJSON just uses the google api client. The main thing is that https://github.com/fog/fog-google/blob/master/lib/fog/storage/google_json/models/files.rb#L34-L41, https://github.com/fog/fog-google/blob/master/lib/fog/storage/google_xml/models/files.rb#L55-L67, and https://github.com/fog/fog-aws/blob/master/lib/fog/aws/models/storage/files.rb#L59-L77 act similarly from an input standpoint. The requests can be wildly different.

**Comment by emilymye** [comment]
@tcdowney Sorry, I just saw the example in the original comment.

@icco I assumed the passed-in block was the format expected by Excon - each of these file.get methods just pass through the block to the underlying request and for AWS/GoogleXML/the previous google-api-client calls, it gets set as the Excon `:response_block`param . The Google API doesn't take a streaming response block so we'd have to give it an IO stream instead or fake it using the buffer file.

**Comment by icco** [comment]
@emilymye Yeah. Can we turn a streaming block and transform it into an IO Stream?

**Comment by emilymye** [comment]
We would basically be creating an IO stream and reading it into the block, and then letting @tcdowney write it back to an IO stream in this case. So, it is POSSIBLE - I just hate it a little. 😭

**Comment by tcdowney** [comment]
Wow thanks for getting a PR out so quickly @emilymye!! We'll try it out on Monday. :+1:

**Comment by icco** [comment]
@tcdowney I merged into master. I'll do a release once you sign off on it.

**Comment by tcdowney** [comment]
@icco 

We're having some difficulty testing `master` of this repo because the method signature of `directories#get` changed here:
https://github.com/fog/fog-google/commit/2902d55a04b178a08181e0871466e730526c3868#diff-38f0b74473bc10007c3bcace35c39e54R12

We call it with the `max_keys` option and it's now raising an ArgumentError:
```
[2018-01-22 21:48:35+0000] ArgumentError: unknown keyword: max_keys
[2018-01-22 21:48:35+0000] /var/vcap/packages/cloud_controller_ng/cloud_controller_ng/vendor/cache/fog-google-e8ecf6a5f0c7/lib/fog/storage/google_json/models/directories.rb:12:in `get'
```

Is it possible for it to keep the same interface as the `GoogleXML` version?

https://github.com/fog/fog-google/blob/e8ecf6a5f0c728b0303b02d6f161e8cc19571ade/lib/fog/storage/google_xml/models/directories.rb#L15

**Comment by icco** [comment]
@tcdowney

**Comment by tcdowney** [comment]
@icco 

Yeah that's interesting... I think for the interface it's helpful if it can accept what the other implementations allow though I agree it's a little unfortunate that it doesn't actually do anything and is silent about it. :man_shrugging: I don't think the Openstack implementation works with `max-keys` either, but it doesn't error and this would be similar to that behavior at least: https://github.com/fog/fog-openstack/blob/a0dff17ba9a0054c665aa967a32be19506693269/lib/fog/storage/openstack/models/directories.rb#L15

**Comment by tcdowney** [comment]
One last thing we've seen so far is that `files#head` was replaced with `files#metadata`:
https://github.com/fog/fog-google/commit/da9619b34274e8112674e1ddced2392f1c5aa4f7#diff-5d9e2ffaf754e4d81109929c33f36ed6R48

I'm not very familiar with the difference between the two implementations, but it seems like we could alias `#head` to it so consumers aren't broken? It seems like the intent behind the methods are still pretty much the same?

Thanks!

**Comment by icco** [comment]
Yup, what you're saying makes sense. My original goal in this was to keep models pretty close, but was okay with changes when proposed mainly because we really have no idea how people use this library. Glad we can make some minor tweaks and keep if working for you. Thanks for reporting @tcdowney.

**Comment by tcdowney** [comment]
@icco we're still seeing:
```
[2018-01-23 17:54:57+0000] NameError: undefined local variable or method `head' for Fog::Storage::GoogleJSON::Files:Class
```

We made the method names symbols inline in our code here and it works:
https://github.com/fog/fog-google/commit/06bb0dae2e707f3bf255cc8e4381e99b039ead36#diff-5d9e2ffaf754e4d81109929c33f36ed6R55

Thanks for your quick responses!

**Comment by tcdowney** [comment]
Submitted a PR for the above issue: https://github.com/fog/fog-google/pull/295

**Comment by icco** [comment]
Awesome, thanks. Merged. I've been meaning to write a bot that assigns all PRs to me... Would solve the missing PRs that don't mention me or @fog/fog-google

**Comment by tcdowney** [comment]
Thanks, can confirm that we can now successfully call `files#get` with a block.

**Comment by icco** [comment]
Awesome. Released 1.1.0 with this fix.
